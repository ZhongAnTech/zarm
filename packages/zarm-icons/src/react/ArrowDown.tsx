import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowDown = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M494.094 733.426a41.472 41.472 0 01-20.023-11.126L140.16 388.388c-16.272-16.271-16.272-42.653 0-58.925 16.271-16.272 42.653-16.272 58.925 0L503.66 634.037 808.186 329.51c16.271-16.272 42.653-16.272 58.925 0s16.272 42.654 0 58.926L533.2 722.347c-10.607 10.607-25.508 14.3-39.106 11.08z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowDown);
export default ForwardRef;
