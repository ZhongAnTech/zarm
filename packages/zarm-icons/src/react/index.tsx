export { default as ArrowDown } from './ArrowDown';
export { default as ArrowLeft } from './ArrowLeft';
export { default as ArrowRight } from './ArrowRight';
export { default as ArrowUp } from './ArrowUp';
export { default as CloseCircleFill } from './CloseCircleFill';
export { default as CloseCircle } from './CloseCircle';
export { default as Close } from './Close';
export { default as DeleteKey } from './DeleteKey';
export { default as Empty } from './Empty';
export { default as Keyboard } from './Keyboard';
export { default as Minus } from './Minus';
export { default as PlusCircle } from './PlusCircle';
export { default as Plus } from './Plus';
export { default as Search } from './Search';
export { default as SuccessCircle } from './SuccessCircle';
export { default as Success } from './Success';
export { default as Volume } from './Volume';
export { default as WarningCircle } from './WarningCircle';
export { default as Warning } from './Warning';
