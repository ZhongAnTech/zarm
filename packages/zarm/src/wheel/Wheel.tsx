import React, { Component } from 'react';
import classnames from 'classnames';
import BScroll from 'better-scroll';
import isEqual from 'lodash/isEqual';
import BaseWheelProps, { WheelItem, WheelValue } from './PropsType';
import { isArray } from '../utils/validate';

const getValue = (props, defaultValue?: any) => {
  if ('value' in props) {
    return props.value;
  }
  if ('defaultValue' in props) {
    return props.defaultValue;
  }
  if (isArray(props.dataSource) && props.dataSource[0]) {
    return props.dataSource[0][props.valueMember];
  }
  return defaultValue;
};

export interface WheelProps extends BaseWheelProps {
  prefixCls?: string;
  className?: string;
}

export default class Wheel extends Component<WheelProps, any> {
  private BScroll;

  private wrapper;

  static defaultProps: WheelProps = {
    prefixCls: 'za-wheel',
    dataSource: [],
    valueMember: 'value',
    itemRender: (item) => item!.label as string,
    stopScroll: false,
  };

  componentDidMount() {
    const { prefixCls, dataSource, disabled } = this.props;
    const value = getValue(this.props);
    const initIndex = this.getSelectedIndex(value, dataSource);
    this.BScroll = new BScroll(this.wrapper, {
      wheel: {
        selectedIndex: initIndex,
        wheelWrapperClass: `${prefixCls}-content`,
        wheelItemClass: `${prefixCls}-item`,
      },
      probeType: 3,
    });

    disabled && this.BScroll.disable();

    this.BScroll.on('scrollEnd', () => {
      this.handleScrollEnd();
    });
  }

  componentDidUpdate(prevProps) {
    const { value, dataSource, disabled, stopScroll } = this.props;
    disabled && this.BScroll.disable();
    if (!isEqual(prevProps.dataSource, dataSource)) {
      this.BScroll.refresh();
    }
    const oldIndex = this.getSelectedIndex(prevProps.value, prevProps.dataSource);
    const newIndex = this.getSelectedIndex(value, dataSource);
    if (newIndex !== oldIndex) {
      this.BScroll.wheelTo(newIndex);
    }

    if (stopScroll && prevProps.stopScroll !== stopScroll) {
      this.BScroll.stop();
    }
  }

  componentWillUnmount() {
    this.BScroll.destroy();
  }

  handleScrollEnd = () => {
    const { dataSource: curDataSource, valueMember } = this.props;
    const index = this.BScroll.getSelectedIndex();
    const child = curDataSource[index];

    if (child) {
      this.fireValueChange(child[valueMember!]);
    }
  };

  getSelectedIndex = (value?: WheelValue, dataSource?: Array<WheelItem>): number => {
    const { valueMember } = this.props;
    let index = 0;
    if (dataSource) {
      dataSource.some((item, i) => {
        if (item[valueMember!] === value) {
          index = i;
          return true;
        }
        return false;
      });
    }
    return index;
  };

  fireValueChange = (value) => {
    const currentValue = getValue(this.props);
    if (value === currentValue) {
      return;
    }

    const { onChange } = this.props;
    if (typeof onChange === 'function') {
      onChange(value);
    }
  };

  render() {
    const { prefixCls, className, valueMember, dataSource, itemRender, disabled } = this.props;
    const value = getValue(this.props);

    const items = dataSource!.map((item, index) => {
      const itemCls = classnames(`${prefixCls}__item`, {
        [`${prefixCls}__item--selected`]: value === item[valueMember!],
        [`${prefixCls}__item--disabled`]: disabled,
      });

      return (
        <div key={+index} className={itemCls}>
          {itemRender!(item)}
        </div>
      );
    });

    const rollerCls = classnames(prefixCls, className);

    return (
      <div
        className={rollerCls}
        ref={(wrapper) => {
          this.wrapper = wrapper;
        }}
      >
        <div className={`${prefixCls}__content`}>{items}</div>
      </div>
    );
  }
}
