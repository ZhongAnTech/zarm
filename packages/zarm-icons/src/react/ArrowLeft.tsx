import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowLeft = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M296.114 508.035c-3.22-13.597.473-28.499 11.079-39.105l333.912-333.912c16.271-16.272 42.653-16.272 58.925 0s16.272 42.654 0 58.926L395.504 498.47l304.574 304.574c16.272 16.272 16.272 42.654 0 58.926s-42.654 16.272-58.926 0L307.241 528.058a41.472 41.472 0 01-11.127-20.023z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowLeft);
export default ForwardRef;
