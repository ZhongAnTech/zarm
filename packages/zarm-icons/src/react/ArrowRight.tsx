import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowRight = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M711.156 508.035a41.472 41.472 0 01-11.126 20.023L366.12 861.97c-16.272 16.272-42.654 16.272-58.926 0s-16.272-42.654 0-58.926L611.767 498.47 307.241 193.944c-16.272-16.272-16.272-42.654 0-58.926 16.271-16.272 42.653-16.272 58.925 0L700.078 468.93c10.606 10.606 14.299 25.508 11.078 39.105z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowRight);
export default ForwardRef;
