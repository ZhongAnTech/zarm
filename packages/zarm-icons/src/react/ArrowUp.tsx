import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowUp = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M513.176 285.05a41.472 41.472 0 0120.024 11.127L867.11 630.089c16.272 16.271 16.272 42.653 0 58.925s-42.654 16.272-58.925 0L503.612 384.44 199.085 688.967c-16.272 16.271-42.654 16.271-58.925 0-16.272-16.272-16.272-42.654 0-58.926L474.07 296.129c10.606-10.606 25.508-14.299 39.105-11.078z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowUp);
export default ForwardRef;
