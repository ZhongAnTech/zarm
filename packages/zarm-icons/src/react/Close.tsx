import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgClose = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M560.149 501.223l255.344 255.344c16.272 16.272 16.272 42.654 0 58.926s-42.654 16.272-58.926 0L501.223 560.149 245.88 815.493c-16.272 16.272-42.654 16.272-58.925 0-16.272-16.272-16.272-42.654 0-58.926l255.344-255.344L186.954 245.88c-16.272-16.272-16.272-42.654 0-58.925 16.271-16.272 42.653-16.272 58.925 0l255.344 255.344 255.344-255.344c16.272-16.272 42.654-16.272 58.926 0 16.272 16.271 16.272 42.653 0 58.925L560.149 501.223z"
          fillRule="nonzero"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgClose);
export default ForwardRef;
