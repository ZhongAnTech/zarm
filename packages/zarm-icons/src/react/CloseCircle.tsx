import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgCloseCircle = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <g fillRule="evenodd">
          <g>
            <path d="M500 460.716l137.493-137.493c10.848-10.848 28.436-10.848 39.284 0l.101.102c10.735 10.848 10.645 28.344-.203 39.08l-138.207 136.78 138.309 138.308c10.848 10.848 10.848 28.436 0 39.284-10.848 10.848-28.436 10.848-39.284 0L498.98 538.264 359.85 675.956c-10.41 10.303-27.188 10.26-37.545-.097-10.34-10.341-10.34-27.107 0-37.448L460.716 500 323.223 362.507c-10.848-10.848-10.848-28.436 0-39.284 10.848-10.848 28.436-10.848 39.284 0L500 460.716z" />
            <path
              d="M190.64 809.36c170.855 170.854 447.865 170.854 618.72 0 170.854-170.855 170.854-447.865 0-618.72-170.855-170.854-447.865-170.854-618.72 0-170.854 170.855-170.854 447.865 0 618.72zm-44.193 44.193c-195.263-195.262-195.263-511.844 0-707.106 195.262-195.263 511.844-195.263 707.106 0 195.263 195.262 195.263 511.844 0 707.106-195.262 195.263-511.844 195.263-707.106 0z"
              fillRule="nonzero"
            />
          </g>
        </g>
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgCloseCircle);
export default ForwardRef;
