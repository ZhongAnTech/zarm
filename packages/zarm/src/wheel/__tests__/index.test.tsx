import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Wheel from '../index';

// function fakeTimers() {
//   performance.timing = {};
//   performance.timing.navigationStart = 0;
// }
// fakeTimers();

describe('Wheel', () => {
  it('Wheel render visible', () => {
    const { container } = render(
      <Wheel
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Wheel set props value', () => {
    const { container } = render(
      <Wheel
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        defaultValue="1"
        value="1"
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Wheel set fieldNames', () => {
    const { container } = render(
      <Wheel
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        fieldNames={{ value: 'value' }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Wheel render defaultValue', () => {
    const { container } = render(
      <Wheel
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        defaultValue="1"
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Wheel disabled', () => {
    const { container } = render(
      <Wheel
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        disabled
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Wheel set props disabled', () => {
    const { container } = render(
      <Wheel
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        defaultValue="1"
        value="1"
        disabled
      />,
    );
    const disabled = container.querySelectorAll('.za-wheel__item--disabled');
    expect(disabled?.length).toBeGreaterThan(0);
  });

  // it('Wheel touch move', () => {
  //   // jest.useFakeTimers();
  //   const { container } = render(
  //     <Wheel
  //       dataSource={[
  //         { value: '1', label: '选项一' },
  //         { value: '2', label: '选项二' },
  //       ]}
  //       defaultValue="1"
  //       value="1"
  //     />
  //   );

  //   // scroll = new BScroll(wrapper.ref('secondRef'), scrollOptions);
  //   // wrapper = scroll.wrapper;
  //   const wheel = container.querySelector('.za-wheel');
  //   fireEvent.touchStart(wheel!, { pointerId: 12, clientY: 0, buttons: 1 });
  //   fireEvent.touchMove(wheel!, { pointerId: 12, clientY: -100, buttons: 1 });
  //   fireEvent.touchMove(wheel!, { pointerId: 12, clientY: -200, buttons: 1 });
  //   fireEvent.touchEnd(wheel!, { pointerId: 12 });

  // });
});
