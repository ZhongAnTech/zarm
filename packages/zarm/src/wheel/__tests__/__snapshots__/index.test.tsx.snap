// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wheel Wheel disabled 1`] = `
<div>
  <div
    class="za-wheel"
  >
    <div
      class="za-wheel__content"
      style="transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-property: transform; transition-duration: 800ms; transform: translateX(0px) translateY(NaNpx);"
    >
      <div
        class="za-wheel__item za-wheel__item--disabled"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项一
      </div>
      <div
        class="za-wheel__item za-wheel__item--disabled"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项二
      </div>
    </div>
  </div>
</div>
`;

exports[`Wheel Wheel render defaultValue 1`] = `
<div>
  <div
    class="za-wheel"
  >
    <div
      class="za-wheel__content"
      style="transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-property: transform; transition-duration: 800ms; transform: translateX(0px) translateY(NaNpx);"
    >
      <div
        class="za-wheel__item za-wheel__item--selected"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项一
      </div>
      <div
        class="za-wheel__item"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项二
      </div>
    </div>
  </div>
</div>
`;

exports[`Wheel Wheel render visible 1`] = `
<div>
  <div
    class="za-wheel"
  >
    <div
      class="za-wheel__content"
      style="transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-property: transform; transition-duration: 800ms; transform: translateX(0px) translateY(NaNpx);"
    >
      <div
        class="za-wheel__item"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项一
      </div>
      <div
        class="za-wheel__item"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项二
      </div>
    </div>
  </div>
</div>
`;

exports[`Wheel Wheel set fieldNames 1`] = `
<div>
  <div
    class="za-wheel"
  >
    <div
      class="za-wheel__content"
      style="transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-property: transform; transition-duration: 800ms; transform: translateX(0px) translateY(NaNpx);"
    >
      <div
        class="za-wheel__item za-wheel__item--selected"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项一
      </div>
      <div
        class="za-wheel__item"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项二
      </div>
    </div>
  </div>
</div>
`;

exports[`Wheel Wheel set props value 1`] = `
<div>
  <div
    class="za-wheel"
  >
    <div
      class="za-wheel__content"
      style="transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-property: transform; transition-duration: 800ms; transform: translateX(0px) translateY(NaNpx);"
    >
      <div
        class="za-wheel__item za-wheel__item--selected"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项一
      </div>
      <div
        class="za-wheel__item"
        style="transform: rotateX(NaNdeg); transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1); transition-duration: 800ms; pointer-events: none;"
      >
        选项二
      </div>
    </div>
  </div>
</div>
`;
